control-plane/agents/core/src/pool/specs.rs

```diff
@@ -0,0 +1,450 @@
+                return Err(SvcError::Conflict {});
+            } else if replica.state.deleted() {
+                return Ok(());
+            }
+            if !destroy_replica {
+                return Err(SvcError::Conflict {});
+            }
+            if !replica.state.deleting() {
+                replica.state = ReplicaSpecState::Deleting;
+                // write it to the store
+                let mut store = registry.store.lock().await;
+                store.put_obj(replica.deref()).await?;
+            }
+            replica.updating = true;
+        }
+
+        if let Some(replica) = replica {
+            let result = node.destroy_replica(request).await;
+            match &result {
+                Ok(_) => {
+                    let mut replica = replica.lock().await;
+                    replica.updating = false;
+                    {
+                        // remove the spec from the persistent store
+                        // if it fails, then fail the request and let the op
+                        // retry
+                        let mut store = registry.store.lock().await;
+                        if let Err(error) =
+                            store.delete_kv(&request.uuid.key()).await
+                        {
+                            if !matches!(error, StoreError::MissingEntry { .. })
+                            {
+                                return Err(error.into());
+                            }
+                        }
+                    }
+                    replica.state = ReplicaSpecState::Deleted;
+                    drop(replica);
+                    // now remove the spec from our list
+                    self.del_replica(&request.uuid).await;
+                }
+                Err(_error) => {
+                    let mut replica = replica.lock().await;
+                    replica.updating = false;
+                }
+            }
+            result
+        } else {
+            node.destroy_replica(&request).await
+        }
+    }
+    pub(super) async fn share_replica(
+        &self,
+        registry: &Registry,
+        request: &ShareReplica,
+    ) -> Result<String, SvcError> {
+        let node = registry.get_node_wrapper(&request.node).await.context(
+            NodeNotFound {
+                node_id: request.node.clone(),
+            },
+        )?;
+
+        if let Some(spec) = self.get_replica(&request.uuid).await {
+            let mut spec = spec.lock().await;
+            if spec.updating {
+                return Err(SvcError::Conflict {});
+            } else if !spec.state.created() {
+                return Err(SvcError::ReplicaNotFound {
+                    replica_id: request.uuid.clone(),
+                });
+            }
+
+            spec.share = request.protocol.into();
+
+            let mut store = registry.store.lock().await;
+            store.put_obj(spec.deref()).await?;
+            node.share_replica(request).await
```

Comment:
Good point, this is where things get hairy. We don't know why that failed, maybe the request simply timed out or the connection broke but it did succeed. What do we want these specs to actually mean, since we're not doing reconciliation on these low level resources?